lib/puppet/util/tagging.rb

```diff
@@ -40,7 +40,7 @@ def tag(*ary)
   # since that results in testing the same string twice
   #
   def tag_if_valid(name)
-    if name.is_a?(String) && name =~ ValidTagRegex
+    if name.is_a?(String) && name =~ ValidTagRegex && Puppet::Resource::MAIN != name
```

Comment:
The only way to find out is probably to run every conceivable test that we have. A PEZ build would be a good start.